File: src/dependencies/crazyflie_hardware/src/crtp/crtp/packers/logging_packer.py
from .toc_packer import TocPacker
from .crtp_packer import CrtpPacker

import struct

from typing import Callable


class LoggingPacker(TocPacker):
    PORT_LOGGING = 5

    # Channels used for the logging port
    TOC_CHANNEL = 0
    CONTROL_CHANNEL = 1
    LOGDATA_CHANNEL = 2

    # Commands used when accessing the Log configurations
    CMD_CREATE_BLOCK = 0
    CMD_APPEND_BLOCK = 1
    CMD_DELETE_BLOCK = 2
    CMD_START_LOGGING = 3
    CMD_STOP_LOGGING = 4
    CMD_RESET_LOGGING = 5
    CMD_CREATE_BLOCK_V2 = 6
    CMD_APPEND_BLOCK_V2 = 7

    def __init__(self, crtp_packer_factory: Callable[[int], CrtpPacker]):
        super().__init__(crtp_packer_factory, self.PORT_LOGGING)

    def _create_block_content(self, content):
        data = struct.pack("<")
        for el in content:
            storage_and_fetch, index = el
            data += struct.pack(
                "<BBB", storage_and_fetch, index & 0xFF, (index >> 8) & 0xFF
            )  # storage and fetch byte
        return data

    def create_block(self, index, content):
        data = struct.pack("<BB", self.CMD_CREATE_BLOCK_V2, index)
        data += self._create_block_content(content)
        return self._prepare_packet(self.CONTROL_CHANNEL, data), True, 2

    def start_block(self, index, period):
        data = struct.pack("<BBB", self.CMD_START_LOGGING, index, period)
        return self._prepare_packet(self.CONTROL_CHANNEL, data), True, 2

    def stop_block(self, index):
        data = struct.pack("<BB", self.CMD_STOP_LOGGING, index)
        return self._prepare_packet(self.CONTROL_CHANNEL, data), True, 2

File: src/dependencies/crazyflie_hardware/src/crtp_cpp/src/crtp_cpp/packer/logging_packer.cpp
#include "crtp_cpp/packer/logging_packer.hpp"
#include <tuple>
#include <cstring> // Add this line


// Commands used when accessing the Log configurations
#define CMD_CREATE_BLOCK 0
#define CMD_APPEND_BLOCK 1
#define CMD_DELETE_BLOCK 2
#define CMD_START_LOGGING 3
#define CMD_STOP_LOGGING 4
#define CMD_RESET_LOGGING 5
#define CMD_CREATE_BLOCK_V2 6
#define CMD_APPEND_BLOCK_V2 7

LoggingPacker::LoggingPacker()
    : TocPacker(PORT_LOGGING) {}


CrtpPacket LoggingPacker::prepare_control_packet(const std::vector<uint8_t>& data)
{
  return CrtpPacker::prepare_packet(CONTROL_CHANNEL, data);
}


CrtpRequest LoggingPacker::create_block(uint8_t block_id, const std::vector<std::pair<uint8_t, uint16_t>>& variables) {
  CrtpRequest request;
  std::vector<uint8_t> data = {CMD_CREATE_BLOCK_V2, block_id};

  std::vector<uint8_t> content_data = create_block_content(variables);
  data.insert(data.end(), content_data.begin(), content_data.end());
  request.packet = prepare_control_packet(data);
  request.expects_response = true;
  request.matching_bytes = 2;
  return request;
}

CrtpRequest LoggingPacker::start_block(uint8_t index, uint8_t period) {
  CrtpRequest request;
  std::vector<uint8_t> data = {CMD_START_LOGGING, index, period};
  request.packet = prepare_control_packet(data);
  request.expects_response = true;
  request.matching_bytes = 2;
  return request;
}

CrtpRequest LoggingPacker::stop_block(uint8_t index) {
  CrtpRequest request;
  std::vector<uint8_t> data = {CMD_STOP_LOGGING, index};
  request.packet = prepare_control_packet(data);
  request.expects_response = true;
  request.matching_bytes = 2;
  return request;
}

CrtpRequest LoggingPacker::reset() {
    CrtpRequest request;

    std::vector<uint8_t> data = {CMD_RESET_LOGGING};
    request.packet = prepare_control_packet(data);
    request.expects_response = true;
    request.matching_bytes = 1;
    return request;
}


std::vector<uint8_t> LoggingPacker::create_block_content(const std::vector<std::pair<uint8_t, uint16_t>>& variables) {
  std::vector<uint8_t> data;
  for (const auto& var : variables) {
    uint8_t type = var.first;
    uint16_t var_id = var.second;

    data.push_back(type);
    data.push_back(static_cast<uint8_t>(var_id & 0xFF));       // Lower byte
    data.push_back(static_cast<uint8_t>((var_id >> 8) & 0xFF)); // Higher byte
  }
  return data;
}
